### DISCUSSION TITLE
What is the suggested way of static data fetching for persistent components?

### DISCUSSION THREAD

**Discussion by josh-unwin** [vercel/next.js#15123] [Help] [answered]
Hi there,

I have a basic question regarding data fetching with getStaticProps etc. Given that it's page based (can't use on components), what is the suggested way to fetch data for components which persist as part of the layout across all pages?

For example, my nav bar. My navbar component is populated with data from my CMS. The `navbar` is loaded within my `layout` HOC, which in turn is loaded in my custom `_app`. I can't fetch the data from the custom app as per [this post](https://github.com/vercel/next.js/discussions/10949), so I'm not sure how else it should be done elegantly?

**Comment by magicspon** [comment]
I'm also interested in a solution for this issue.

**Comment by goncy** [comment]
_app has access to pageProps, I think you might fetch the props you need on getStaticProps and use them on _app through component props right?

Another solution would be to move the navbar component to every page (which I don't think its necessary with the firest approach)

**Comment by magicspon** [comment]
@goncy The issue with getting the data from pageProps is you'd have to include your global queries on every page.

So rather than requesting the header/footer stuff once, in one place, you'd need to request it in every single page. eeee!

**Comment by jserrao** [comment]
Yeah, ran into this exact same problem @josh-unwin. It seems like there is a missing piece to the framework. 

If you do a custom call up in a custom version of `_app` inside of pages (which is where this would logically go), the whole site [goes SSR instead of static generation](https://nextjs.org/docs/advanced-features/custom-app#caveats).

**Reply by jserrao** [reply]
I solved this more or less, by using Apollo. The data doesn't get fetched server-side but rather through the client using an Apollo provider. I wrote up the whole approach on Stack Overflow if you're interested @josh-unwin, @magicspon & @goncy (would love your feedback):
https://stackoverflow.com/questions/63494479/best-practice-for-next-js-data-fetching-inside-a-component

**Reply by josh-unwin** [reply]
@jserrao I'm using Apollo all over my app too, but Apollo isn't static generation is it? Still believe there's quite a significant shortcoming here.

**Comment by goncy** [comment]
You can still have some kind of cache on your fetch util if you "really need" to do that, something like:
```js
const cache = new Map();

const fetchFromCache = async (id, ...params) => {
  return cache.get(id) || fetch(...params);
}
```
And from your pages just do something like:
```js
fetch(`myRequestCalledLikeThis`, /* your fetch util params */)
```

**Comment by josh-unwin** [comment]
Bumping this question, I'm still not sure how to statically fetch data for a component that is persistent across every/many pages without having to request in every individual page.

**Comment by balazsorban44** [comment, marked as answer]
This is a re-appearing discussion, I think this one is the most "official" one, that could solve the problem:
https://github.com/vercel/next.js/discussions/10949

**Reply by josh-unwin** [reply]
> This is a re-appearing discussion, I think this one is the most "official" one, that could solve the problem:
> #10949

You're quite right, will mark this as the answer to point people there.

**Reply by balazsorban44** [reply]
It would be so great if people first looked up if their issue was in discussion already before opening new ones, both the Next core team and users' focus could be gathered in one place and maybe because of all the duplicates that team thinks that there is not enough interest for this feature, when in fact there are so many requesting it just in different places. I guess github could do a better job displaying the users that their request has already been asked.

**Reply by josh-unwin** [reply]
In my defence I actually did search and link to that exact post in my OP, created this one because I wasn't entirely sure if there wasn't an alternative solution.

**Reply by mitchierichie** [reply]
Could anyone please help with maybe a link to what the actual answer is instead of just the link to a very long discussion? I have spent a bit of time on that thread already and I still don't have an answer to this problem.

**Reply by shadowprompt** [reply]
So, **WE SHOULD GIVE UP** this purpose.
